### PR TITLE
Rework heart useful calculation

### PR DESCRIPTION
**Pull request by mattman107**
If relevant tricks are disabled, and you have more hearts than the minimum require to get everywhere, just make all the hearts useful.

If the player has less starting hearts than required, calculate how many pieces and containers should be converted to useful. Had to do some shenanigans to make it work with all the item pool options. In my testing (starting 1 and 2 hearts and all the item pool options) it seemed to work out. 

Would be good to have someone double check my math as I clearly need it from the state of this before.

### CODE REVIEW

**Comment by mattman107** [comment]
Drafting this as I want to add some unit tests while I'm here too.

File: worlds/oot_soh/ItemPool.py [hunk 1]
```diff
@@ -498,27 +498,36 @@ def create_item_pool(world: "SohWorld") -> None:
...
+    if not (world.options.enable_all_tricks or str(Tricks.FEWER_TUNIC_REQUIREMENTS) in world.options.tricks_in_logic.value):
+        min_hearts_needed: int = 3
+        if world.multiworld.state.soh_heart_count[world.player] >= min_hearts_needed:
+            items_to_create[Items.PIECE_OF_HEART] -= create_special_progression_item(world, Items.PIECE_OF_HEART, ItemClassification.useful, items_to_create[Items.PIECE_OF_HEART])
+            items_to_create[Items.PIECE_OF_HEART_WINNER] -= create_special_progression_item(world, Items.PIECE_OF_HEART_WINNER, ItemClassification.useful, items_to_create[Items.PIECE_OF_HEART_WINNER])
+            items_to_create[Items.HEART_CONTAINER] -= create_special_progression_item(world, Items.HEART_CONTAINER, ItemClassification.useful, items_to_create[Items.HEART_CONTAINER])
+        else:
+            # starting health is less than min_hearts_needed
+            # Get the total number of hearts that don't need to be progression
+            total_non_progression_hearts: int = min_hearts_needed - world.multiworld.state.soh_heart_count[world.player]
```
Comment (jeromkiller) [inline worlds/oot_soh/ItemPool.py]: I think you're calculating the amount of hearts that should be progressive here.
don't you want to do `20 - min_hearts_needed - world.multiworld.state.soh_heart_count[world.player]` instead?

File: worlds/oot_soh/ItemPool.py [hunk 2]
```diff
@@ -498,27 +498,36 @@ def create_item_pool(world: "SohWorld") -> None:
...
+                    non_progression_piece_amount = 0
+                elif world.options.item_pool == "scarce":
+                    non_progression_piece_amount = min(non_progression_piece_amount , temp_non_progression_hearts) if world.options.item_pool != "minimal" or non_progression_container_amount > 0 else 0
+
+                if non_progression_piece_amount > 0:
+                    items_to_create[Items.PIECE_OF_HEART] -= create_special_progression_item(world, Items.PIECE_OF_HEART, ItemClassification.useful, ((non_progression_piece_amount * 4) - 1))
+                    items_to_create[Items.PIECE_OF_HEART_WINNER] -= create_special_progression_item(world, Items.PIECE_OF_HEART_WINNER, ItemClassification.useful, 1)
+
+                # Try to convert the containers second
+                if total_non_progression_hearts <= non_progression_container_amount:
```
Comment (jeromkiller) [inline worlds/oot_soh/ItemPool.py]: the value of total_non_progressive_hearts doesn't get updated anywhere. If we remove some of the heart containers the total_non_progression hearts should be lowered right?

File: worlds/oot_soh/ItemPool.py [hunk 3]
```diff
@@ -498,27 +498,36 @@ def create_item_pool(world: "SohWorld") -> None:
...
+        if world.multiworld.state.soh_heart_count[world.player] >= min_hearts_needed:
+            items_to_create[Items.PIECE_OF_HEART] -= create_special_progression_item(world, Items.PIECE_OF_HEART, ItemClassification.useful, items_to_create[Items.PIECE_OF_HEART])
+            items_to_create[Items.PIECE_OF_HEART_WINNER] -= create_special_progression_item(world, Items.PIECE_OF_HEART_WINNER, ItemClassification.useful, items_to_create[Items.PIECE_OF_HEART_WINNER])
+            items_to_create[Items.HEART_CONTAINER] -= create_special_progression_item(world, Items.HEART_CONTAINER, ItemClassification.useful, items_to_create[Items.HEART_CONTAINER])
+        else:
+            # starting health is less than min_hearts_needed
+            # Get the total number of hearts that don't need to be progression
+            total_non_progression_hearts: int = min_hearts_needed - world.multiworld.state.soh_heart_count[world.player] 
+            # Container amount to make useful
+            non_progression_container_amount: int = items_to_create[Items.HEART_CONTAINER]
```
Comment (jeromkiller) [inline worlds/oot_soh/ItemPool.py]: I think you're getting a little hung up on calling these non_progression_*, from the looks of it you're just getting the total count but you're not modifying the value further to subtract the mount of hearts that would be progressive from them.

Wouldn't it be clearer if this was just called heart_container_amount?

File: worlds/oot_soh/ItemPool.py [hunk 4]
```diff
@@ -498,27 +498,36 @@ def create_item_pool(world: "SohWorld") -> None:
...
+            # Heart Piece amount to make useful. Convert to regular heart here then convert back later for simplicity
+            non_progression_piece_amount: int = int((items_to_create[Items.PIECE_OF_HEART] + items_to_create[Items.PIECE_OF_HEART_WINNER]) / 4)
```
Comment (jeromkiller) [inline worlds/oot_soh/ItemPool.py]: Same for non_progression_piece amount, you are doing some math on that but you're storing that in a variable called temp_*

**Review by jeromkiller — CHANGES_REQUESTED** [review]
It's a little early in the morning for me so maybe I'm just going cross eyed, but I think the math doesn't check out yet. 
By the looks of things we're putting the amount of progression hearts we need to add in `total_non_progression_hearts`.

You should also look into adding the unit test and changes to `create_special_progression_item` from #266 to this PR so we have a test to see if we have enough hearts after collecting them all. as well as a fix to the edge case of creating a negative amount of special progression items

File: worlds/oot_soh/ItemPool.py (hunk 1 above)
Comment (mattman107) [reply worlds/oot_soh/ItemPool.py]: I thought on this and we are both a bit wrong.

I think technically the formula we need is:
`(max_hearts - starting hearts) - (min_hearts_needed - starting hearts)`

To take that to a real example:
`(20 - 2) - (3 - 2) = 17` 

So there should be 20 hearts total (assume we start on balanced item pool). The minimum number of hearts is always 3 and the player started with 2 hearts. So 1 heart in the pool needs to be progression and the rest (17) can be useful.

This can be simplified to `max hearts - min hearts needed` -> `20 - 3 = 17`. The 2s cancel each other out.

**Comment by mattman107** [comment]
I'm going to mark all of your comments as resolved as the calculation has pretty much been rewritten bottom up. So I don't think any of these apply anymore. Feel free to unresolve them if you think something still applies.

File: worlds/oot_soh/ItemPool.py [hunk 5]
```diff
@@ -498,27 +498,39 @@ def create_item_pool(world: "SohWorld") -> None:
...
+        else:
+             # starting health is less than min_hearts_needed
+            # Get the total number of heart classifications we have to work with
+            total_non_progression_hearts: int = max_hearts - min_hearts_needed
+
+            # Heart Piece amount to make useful. Convert to actual hearts, then convert back later
+            non_progression_piece_amount: int = int((items_to_create[Items.PIECE_OF_HEART] + items_to_create[Items.PIECE_OF_HEART_WINNER]) / 4)
+            
+            if total_non_progression_hearts >= non_progression_piece_amount: 
+                items_to_create[Items.PIECE_OF_HEART] -= create_special_progression_item(world, Items.PIECE_OF_HEART, ItemClassification.useful, ((non_progression_piece_amount * 4) - 1))
```
Comment (jeromkiller) [inline worlds/oot_soh/ItemPool.py]: you should probably check if `non_progression_piece_amount` is greater than zero.
If I'm following the logic correctly when both `total_non_progression_hearts` and `non_progression_piece_amount` are zero, you'll add 0 non_progresion heart pieces but still add 1 non progression winner piece of heart

File: worlds/oot_soh/ItemPool.py [hunk 6]
```diff
@@ -498,27 +498,39 @@ def create_item_pool(world: "SohWorld") -> None:
...
+            items_to_create[Items.HEART_CONTAINER] -= create_special_progression_item(world, Items.HEART_CONTAINER, ItemClassification.useful, items_to_create[Items.HEART_CONTAINER])
+        else:
+             # starting health is less than min_hearts_needed
+            # Get the total number of heart classifications we have to work with
+            total_non_progression_hearts: int = max_hearts - min_hearts_needed
+
+            # Heart Piece amount to make useful. Convert to actual hearts, then convert back later
+            non_progression_piece_amount: int = int((items_to_create[Items.PIECE_OF_HEART] + items_to_create[Items.PIECE_OF_HEART_WINNER]) / 4)
+            
+            if total_non_progression_hearts >= non_progression_piece_amount:
```
Comment (jeromkiller) [inline worlds/oot_soh/ItemPool.py]:
```suggestion
            if (total_non_progression_hearts >= non_progression_piece_amount)
                and (non_progression_piece_amount > 0): 
```

**Review by jeromkiller — CHANGES_REQUESTED** [review]
Looks good, one minor suggestion I'd like to see as extra insurance

File: worlds/oot_soh/ItemPool.py (hunk 5 above)
Comment (mattman107) [reply worlds/oot_soh/ItemPool.py]: Is there ever a scenario that could happen?

If [starting hearts is ever less than max hearts](https://github.com/HarbourMasters/Archipelago-SoH/blob/0bb0aeb76dc538e2707b9f331966d0b6be63659e/worlds/oot_soh/ItemPool.py#L303), there are guaranteed 3 heart pieces and 1 winner piece in the pool. 

I don't mind adding the safety, I just don't think that could ever happen though. LMK if you are seeing a specific example of that being possible.

**Review by jeromkiller — APPROVED** [review]
LGTM
